### PR TITLE
Add timezone scalar function

### PR DESCRIPTION
**Pull request by marregui**
## Summary of the changes / Why this improves CrateDB

This is a draft, to check that I understood the requirements so far and to continue
the specification of the At TIME ZONE parser syntax, which will be built atop

## Checklist

 - [x] User relevant changes are recorded in ``CHANGES.txt``
 - [x] Touched code is covered by tests
 - [x] Documentation has been updated if necessary
 - [x] [CLA](https://crate.io/community/contribute/cla/) is signed
 - [x] This does not contain breaking changes, or if it does:
    - It is released within a major release
    - It is recorded in ``CHANGES.txt``
    - It was marked as deprecated in an earlier release if possible
    - You've thought about the consequences and other components are adapted
      (E.g. AdminUI)

### CODE REVIEW

**Comment by marregui** [comment]
I believe the tests will fail when we transition to DST, I am looking into it.

**Comment by marregui** [comment]
The function outputs in ISO8601 which I am not certain it is acceptable. 
I notice that DateFormatFunction formats by default as in:
DEFAULT_FORMAT = "%Y-%m-%dT%H:%i:%s.%fZ";
with no timezone, which is similar, EXCEPT in the rounding of the fraction of seconds

File: blackbox/docs/appendices/release-notes/unreleased.rst [hunk 1]
```diff
@@ -89,6 +89,8 @@ Changes
 
 - Added left and right scalar functions.
 
+- Added timezone scalar function.
```
Comment (mfussenegger) [inline blackbox/docs/appendices/release-notes/unreleased.rst]: Could you add a reference to the relevant docs using `:ref:...` ?

File: blackbox/docs/general/builtins/scalar.rst [hunk 2]
```diff
@@ -734,6 +734,57 @@ The ``timezone`` will be ``UTC`` if not provided::
...
+
+.. csv-table::
+   :header: "Type of timestamp", "Return Type", "Description"
+
+   timestamp without time zone OR bigint, timestamp with time zone, Treat given timestamp without time zone as located in the specified timezone
+   timestamp with time zone, timestamp without time zone, Convert given timestamp with time zone to the new timezone with no time zone designation
+
+::
+
+    cr> select timezone('Europe/Madrid', 257504400000) as epoque;
```
Comment (mfussenegger) [inline blackbox/docs/general/builtins/scalar.rst]: Maybe we could change those to also include a formatted variant.

```
select timezone('Europe/Madrid', 257504400000) as epoque,
          date_format('%Y-%m-%d %h:%i', timezone(...
```

Given that the time in ms is hard to parse.

File: blackbox/docs/general/builtins/scalar.rst [hunk 3]
```diff
@@ -734,6 +734,57 @@ The ``timezone`` will be ``UTC`` if not provided::
...
+::
+
+    TIMEZONE(timezone, timestamp)
+
+It has two variants depending on the type of ``timestamp``:
+
+.. csv-table::
+   :header: "Type of timestamp", "Return Type", "Description"
+
+   timestamp without time zone OR bigint, timestamp with time zone, Treat given timestamp without time zone as located in the specified timezone
```
Comment (mfussenegger) [inline blackbox/docs/general/builtins/scalar.rst]: We tend to hard-wrap text by 80 cols

File: sql/src/main/java/io/crate/expression/scalar/timestamp/TimezoneFunction.java [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+                            "invalid argument to %s: expected second parameter" +
+                            " to be of either type [%s, %s, %s] but found %s",
+                            NAME,
+                            LongType.INSTANCE.getName(),
+                            TimestampType.INSTANCE_WITH_TZ.getName(),
+                            TimestampType.INSTANCE_WITHOUT_TZ.getName(),
+                            signatureTypes.get(1)));
+                }
+                return new TimezoneFunction(new FunctionInfo(new FunctionIdent(NAME, signatureTypes), returnType),
+                                            paramHadTimezone);
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/expression/scalar/timestamp/TimezoneFunction.java]: Would it make sense to infer this based on the `FunctionInfo` return type, instead of adding an extra parameter?

File: sql/src/main/java/io/crate/expression/scalar/timestamp/TimezoneFunction.java [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+    }
+
+    @Override
+    public Long evaluate(TransactionContext txnCtx, Input<Object>... args) {
+        assert args.length == 2 : String.format(Locale.ENGLISH,
+                                                "number of arguments must be 2, got %d instead",
+                                                args.length);
+        String zoneStr = (String) args[0].value();
+        if (zoneStr == null || zoneStr.isBlank()) {
+            throw new IllegalArgumentException(String.format(Locale.ENGLISH,
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/expression/scalar/timestamp/TimezoneFunction.java]: In PostgreSQL a zone with `null` as value results in `null` as result

File: sql/src/main/java/io/crate/expression/scalar/timestamp/TimezoneFunction.java [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+        try {
+            zoneId = ZoneId.of(zoneStr);
+        } catch (DateTimeException e) {
+            throw new IllegalArgumentException(String.format(Locale.ENGLISH,
+                                                             "invalid argument to %s: %s",
+                                                             NAME, zoneStr));
+        }
+        Number utcTimestamp = (Number) args[1].value();
+        if (utcTimestamp == null) {
+            throw new IllegalArgumentException(String.format(Locale.ENGLISH,
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/expression/scalar/timestamp/TimezoneFunction.java]: Same here, should result in `null` as result.

File: sql/src/main/java/io/crate/expression/scalar/timestamp/TimezoneFunction.java [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+import java.time.ZoneId;
+import java.time.ZonedDateTime;
+import java.util.List;
+import java.util.Locale;
+
+public class TimezoneFunction extends Scalar<Long, Object> {
+
+    public static final String NAME = "timezone";
+    private static final ZoneId UTC = ZoneId.of("UTC");
+    private static final FuncParams FUNCTION_PARAMS = FuncParams.builder(Param.STRING, Param.ANY).build();
```
Comment (kovrus) [inline sql/src/main/java/io/crate/expression/scalar/timestamp/TimezoneFunction.java]: the second function argument shall not be any, right?

File: sql/src/main/java/io/crate/expression/scalar/timestamp/TimezoneFunction.java (hunk 4 above)
Comment (marregui) [reply sql/src/main/java/io/crate/expression/scalar/timestamp/TimezoneFunction.java]: You are absolutely right, I made this mistake because my brain was running low on stamina.

File: blackbox/docs/appendices/release-notes/unreleased.rst (hunk 1 above)
Comment (seut) [reply blackbox/docs/appendices/release-notes/unreleased.rst]: Please add new entries always on top of the list.

File: sql/src/main/java/io/crate/expression/scalar/timestamp/TimezoneFunction.java (hunk 5 above)
Comment (marregui) [reply sql/src/main/java/io/crate/expression/scalar/timestamp/TimezoneFunction.java]: not sure what it returns, it is rendered as blanc, however the return type is a timestamp with/without time zone (depending on the type of the second parameter). I will return null, and NULL will be rendered.

Comment (mfussenegger) [reply sql/src/main/java/io/crate/expression/scalar/timestamp/TimezoneFunction.java]: You can do something like `\pset null '💩'` in psql to see `null` values in a different format. (Can also be added to `.psqlrc` to have it permanently set.

`zoneStr == null`  should result in `null`

File: sql/src/main/java/io/crate/expression/scalar/timestamp/TimezoneFunction.java (hunk 7 above)
Comment (marregui) [reply sql/src/main/java/io/crate/expression/scalar/timestamp/TimezoneFunction.java]: Hi, it can be either of 'timestamp with time zone', 'timestamp without time zone' or bigint, so I take ANY and check the actual type dynamically.

File: sql/src/main/java/io/crate/expression/scalar/timestamp/TimezoneFunction.java (hunk 5 above)
Comment (marregui) [reply sql/src/main/java/io/crate/expression/scalar/timestamp/TimezoneFunction.java]: alles klar, same for the comment on the timestamp param being null, thank you

File: blackbox/docs/appendices/release-notes/unreleased.rst (hunk 1 above)
Comment (marregui) [reply blackbox/docs/appendices/release-notes/unreleased.rst]: Yes and of course.

File: blackbox/docs/general/builtins/scalar.rst (hunk 2 above)
Comment (marregui) [reply blackbox/docs/general/builtins/scalar.rst]: done

File: blackbox/docs/general/builtins/scalar.rst (hunk 3 above)
Comment (marregui) [reply blackbox/docs/general/builtins/scalar.rst]: done

File: sql/src/main/java/io/crate/expression/scalar/timestamp/TimezoneFunction.java (hunk 7 above)
Comment (kovrus) [reply sql/src/main/java/io/crate/expression/scalar/timestamp/TimezoneFunction.java]: would something like that make more sense?
```
FuncParams.builder(Param.STRING, of(DataTypes.TIMESTAMPZ, DataTypes.TIMESTAMP)
```

Comment (marregui) [reply sql/src/main/java/io/crate/expression/scalar/timestamp/TimezoneFunction.java]: mmmm on it

Comment (marregui) [reply sql/src/main/java/io/crate/expression/scalar/timestamp/TimezoneFunction.java]: ok, yes, more restrictive signature, done

**Comment by marregui** [comment]
I have squashed

File: blackbox/docs/general/builtins/scalar.rst [hunk 8]
```diff
@@ -706,9 +706,12 @@ If no ``format_string`` is given the default format will be used::
     +-----------------------------+
     SELECT 1 row in set (... sec)
 
+.. _scalar-timezone:
```
Comment (mfussenegger) [inline blackbox/docs/general/builtins/scalar.rst]: Should the link for the `timezone` scalar be further down? This section here is the description of `timezone` in the context of `date_format`

File: sql/src/main/java/io/crate/expression/scalar/timestamp/TimezoneFunction.java [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+            @Override
+            public FunctionImplementation getForTypes(List<DataType> signatureTypes) throws IllegalArgumentException {
+                DataType returnType = null;
+                int timestampParamTypeId = signatureTypes.get(1).id();
+                switch (timestampParamTypeId) {
+                    case LongType.ID:
+                    case TimestampType.ID_WITHOUT_TZ:
+                        returnType = DataTypes.TIMESTAMPZ;
+                        break;
+                    case UndefinedType.ID:
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/expression/scalar/timestamp/TimezoneFunction.java]: Why is `undefined` allowed?

File: sql/src/main/java/io/crate/expression/scalar/timestamp/TimezoneFunction.java [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+    private static final FuncParams FUNCTION_PARAMS = FuncParams.builder(Param.STRING,
+                                                                         Param.of(DataTypes.TIMESTAMPZ,
+                                                                                  DataTypes.TIMESTAMP,
+                                                                                  DataTypes.LONG)).build();
+
+    public static void register(ScalarFunctionModule module) {
+        module.register(NAME, new BaseFunctionResolver(FUNCTION_PARAMS) {
+            @Override
+            public FunctionImplementation getForTypes(List<DataType> signatureTypes) throws IllegalArgumentException {
+                DataType returnType = null;
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/expression/scalar/timestamp/TimezoneFunction.java]: I think the `returnType = null` assignment is redundant, since it will always be assigned.

File: sql/src/main/java/io/crate/expression/scalar/timestamp/TimezoneFunction.java [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+    @Override
+    public Long evaluate(TransactionContext txnCtx, Input<Object>... args) {
+        assert args.length == 2 : String.format(Locale.ENGLISH,
+                                                "number of arguments must be 2, got %d instead",
+                                                args.length);
+        String zoneStr = (String) args[0].value();
+        if (zoneStr == null) {
+            return null;
+        }
+        ZoneId zoneId = null;
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/expression/scalar/timestamp/TimezoneFunction.java]: Same here, the assignment to `null` looks redundant.

File: sql/src/main/java/io/crate/expression/scalar/timestamp/TimezoneFunction.java [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+                                                             " time zone \"%s\" not recognized",
+                                                             zoneStr));
+        }
+        Number utcTimestamp = (Number) args[1].value();
+        if (utcTimestamp == null) {
+            return null;
+        }
+
+        Instant instant = Instant.ofEpochMilli(utcTimestamp.longValue());
+        boolean paramHadTimezone = info.returnType() == DataTypes.TIMESTAMP;
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/expression/scalar/timestamp/TimezoneFunction.java]: Isn't `TIMESTAMP` the one without time zone?

File: sql/src/main/java/io/crate/expression/scalar/timestamp/TimezoneFunction.java (hunk 9 above)
Comment (marregui) [reply sql/src/main/java/io/crate/expression/scalar/timestamp/TimezoneFunction.java]: the type of null when null is typed as an input parameter

Comment (mfussenegger) [reply sql/src/main/java/io/crate/expression/scalar/timestamp/TimezoneFunction.java]: I think that shouldn't be necessary, also long shouldn't be  part of the Param definition. 

The user provided values will then be cast to the signature's types

File: blackbox/docs/general/builtins/scalar.rst (hunk 8 above)
Comment (marregui) [reply blackbox/docs/general/builtins/scalar.rst]: done

File: sql/src/main/java/io/crate/expression/scalar/timestamp/TimezoneFunction.java (hunk 10 above)
Comment (marregui) [reply sql/src/main/java/io/crate/expression/scalar/timestamp/TimezoneFunction.java]: done

File: sql/src/main/java/io/crate/expression/scalar/timestamp/TimezoneFunction.java (hunk 11 above)
Comment (marregui) [reply sql/src/main/java/io/crate/expression/scalar/timestamp/TimezoneFunction.java]: done

File: sql/src/main/java/io/crate/expression/scalar/timestamp/TimezoneFunction.java (hunk 12 above)
Comment (marregui) [reply sql/src/main/java/io/crate/expression/scalar/timestamp/TimezoneFunction.java]: Yes, if the return parameter has no time zone, it means that the input parameter 'timestamp'  does have time zone and vice versa. Originally I went for the extra parameter to the constructor to make it explicit and avoid this confusion.

File: sql/src/main/java/io/crate/expression/scalar/timestamp/TimezoneFunction.java (hunk 9 above)
Comment (marregui) [reply sql/src/main/java/io/crate/expression/scalar/timestamp/TimezoneFunction.java]: done, the core is now simpler, thank you for this

File: blackbox/docs/general/builtins/scalar.rst [hunk 13]
```diff
@@ -706,9 +706,11 @@ If no ``format_string`` is given the default format will be used::
     +-----------------------------+
     SELECT 1 row in set (... sec)
 
+
 Timezone
 ........
 
+.. _scalar-timezone:
```
Comment (mfussenegger) [inline blackbox/docs/general/builtins/scalar.rst]: I think this should go to line 738 https://github.com/crate/crate/pull/8959/files#diff-c1428c761c17d1bb4ed67b8e27361a96R738

Comment (marregui) [reply blackbox/docs/general/builtins/scalar.rst]: Ok, it makes more sense as it is the explicit place where the function is defined

**Comment by marregui** [comment]
I will use the branch to push the AT TIME ZONE parser feature
